### ISSUE TITLE
TCH004 false positive with boto3 stubs - bedrock-runtime

### ISSUE THREAD

**Issue by sansmoraxz** [astral-sh/ruff#8024]
Using the following python code

```py
import typing

import boto3

from .const import REGION

if typing.TYPE_CHECKING:
    from mypy_boto3_bedrock_runtime import BedrockRuntimeClient   # TCH004 flags here for some reason
    from mypy_boto3_s3 import S3Client
else:
    BedrockClient = boto3.client
    S3Client = boto3.client

def get_bedrock_client() -> BedrockRuntimeClient:
    """Return bedrock client."""
    from botocore.config import Config
    config = Config(
    retries = {
            'max_attempts': 100,
            'mode': 'adaptive',
        },
    )
    return boto3.client(
        service_name='bedrock-runtime',
        region_name=REGION,
        config=config,
        endpoint_url=f'https://bedrock.{REGION}.amazonaws.com')

def get_s3_client() -> S3Client: # doesn't flag here
    """Return S3 client."""
    return boto3.client(
        service_name='s3',
        region_name=REGION)
```

Dependencies:-

```toml
[tool.poetry.group.dev.dependencies]
boto3 = "^1.28.64"
botocore = "^1.31.64"
boto3-stubs = {extras = ["bedrock", "bedrock-runtime", "s3"], version = "^1.28.65"}
ruff = "^0.1.0"
```
Wierdly enough it only flags for `bedrock-runtime`. Not `s3` and `bedrock` client.

**Comment by charliermarsh** [comment]
I believe `BedrockRuntimeClient` _does_ need to be imported here. If you add `from __future__ import annotations`, it doesn't, but without `from __future__ import annotations`, Python _does_ need runtime access to return type annotations on functions. I would expect this code to fail at runtime -- does it not?

(Should `BedrockClient = boto3.client` be `BedrockRuntimeClient = boto3.client`?)

**Comment by sansmoraxz** [comment]
All the mypy  imports here are for generating stubs.  https://pypi.org/project/boto3-stubs/ Specifically `boto3.client` builds and retuns a different type of client for each instance.

**Comment by sansmoraxz** [comment]
Anyway I figured it out. it should've been `BedrockRuntimeClient` in the else part. Forgot to change that when migrating to newer boto3.

**Comment by charliermarsh** [comment]
Thanks, that was my second question, in parentheses!
